BJS2-77432 |  Buying premium access

Raidanmark opened the pull request:
Key Changes:

- Implemented PremiumController with a buyPremium(Long userId, Integer durationDays) endpoint to handle incoming purchase requests.
- Created PremiumPeriod enum to define subscription tiers (1-month, 3-month, 1-year) with corresponding days and prices, including a fromDays() method for period lookup.

- Developed PremiumService containing the core buyPremium(long userId, int durationDays) logic:

1. Checks if the user already has an active premium subscription.
2. Prepares the premium purchase as a Payable item.
3. Invokes TransactionService.buyItem() to process the purchase.
4. Upon successful payment confirmation from TransactionService (which internally uses PaymentService and a payment system client), the user's premium status is updated via PremiumRepository.
5. Returns a TransactionResultDto.


- Added PremiumMapper (MapStruct) for DTO-entity conversion.
- Ensured that the premium purchase flow adheres to the new transaction model, where TransactionService orchestrates payment processing with PaymentService and an external payment gateway client.
- Covered the new functionality with unit tests.

Raidanmark committed:
functional created

Raidanmark committed:
Merge branch 'hydra-master-stream10' into BJS2-77432

Raidanmark committed:
replaced utils in utils package

Raidanmark committed:
Merge remote-tracking branch 'origin/BJS2-77432' into BJS2-77432

danbg12 commented on src/main/java/school/faang/user_service/entity/transaction/Transaction.java:
```diff
@@ -0,0 +1,79 @@
+@AllArgsConstructor
+@Entity
+@Table(name = "Transactions")
+public class Transaction {
```
Убери лишние пробелы везде ) чтобы было все по красоте :D

danbg12 commented on src/main/java/school/faang/user_service/service/utils/PremiumServiceUtils.java:
```diff
@@ -0,0 +1,40 @@
+        User user = userService.getUserById(userId);
+        if (user.getPremium() == null
+        || LocalDateTime.now().isAfter(user.getPremium().getEndDate())) {
+            throw new RuntimeException("User with id " + userId + " has premium");
```
Более понятно будет еслт заменить RunTimeException на боллее узко направленныи ексепшен ну или создать кастомныи типо UserDataException ну или на подобие .

danbg12 commented on src/test/java/school/faang/user_service/service/transaction/TransactionServiceTest.java:
```diff
@@ -0,0 +1,65 @@
+    private PaymentService paymentService;
+    @Mock
+    private TransactionMapper transactionMapper;
+    @InjectMocks
```
Маппер лучше сделать @Spy чтобы у тебя были реальные рабочие методы . Выглядеть это должно так : @Spy  TransactionMapperImpl transactionMapper;

danbg12 commented on src/test/java/school/faang/user_service/service/transaction/TransactionServiceTest.java:
```diff
@@ -0,0 +1,65 @@
+
+        when(transactionServiceUtils.buildTransaction(userId, mockItem)).thenReturn(mockTransaction);
+        when(transactionRepository.save(any())).thenReturn(mockTransaction);
+        when(transactionMapper.toPaymentRequestDto(any())).thenReturn(mockPaymentRequestDto);
```
Мокать маппер плохая идея если помнишь )

danbg12 reviewed (commented):
Претензии нет ) В основном все хорошо!

AlexeyAntipin commented on src/main/java/school/faang/user_service/controller/event/EventController.java:
```diff
@@ -57,7 +60,7 @@ public EventDto updateEvent(@Valid @RequestBody EventDto eventDto) {
 
     @GetMapping(value = "/owned/{id}")
-    public List<EventDto> getOwnedEvents(@PathVariable(value = "id", required = true) Long userId) {
+    public List<EventDto> getOwnedEvents(@PathVariable(value = "id") @NotNull @Positive Long userId) {
```
Можешь добавить в PathVariable параметр required, он сразу будет понятную ошибку выкидывать

Raidanmark replied:
Не очень понял что поменять:
- Если поменять вернуть как было required = true, то он вроде по дефолту такой. Или имеется ввиду что по скольку он дефолный то моно убрать NotNull потому  что о избыточный?
- Если ту аннотацию Required что я делал для валидации то я бы наоборот хотел бы от нее отойти

AlexeyAntipin commented on src/main/java/school/faang/user_service/controller/premium/PremiumController.java:
```diff
@@ -0,0 +1,26 @@
+    private final PremiumService premiumService;
+
+    @PostMapping
+    public TransactionResultDto buyPremium(@RequestParam @NotNull @Positive Long userId,
```
То же самое, в RequestParam можно добавить параметр required

Raidanmark replied:
Не очень понял что поменять:

Если поменять вернуть как было required = true, то он вроде по дефолту такой. Или имеется ввиду что по скольку он дефолный то моно убрать NotNull потому что о избыточный?
Если ту аннотацию Required что я делал для валидации то я бы наоборот хотел бы от нее отойти

AlexeyAntipin replied:
Вообще я имел ввиду первый вариант, но сейчас еще поизучал, required поддерживает null, поэтому оставляй так

AlexeyAntipin commented on src/main/java/school/faang/user_service/entity/premium/PremiumPeriod.java:
```diff
@@ -0,0 +1,31 @@
+
+    public static PremiumPeriod fromDays(Integer days) {
+      return Arrays.stream(values())
+              .filter(period -> period.getDays().equals(days))
```
Лучше Objects.equals

AlexeyAntipin commented on src/main/java/school/faang/user_service/repository/UserRepository.java:
```diff
@@ -26,4 +26,6 @@ WHERE up.end_date > NOW()
     Stream<User> findPremiumUsers();
 
     List<User> findByUsernameLike(String username);
+
```
Лишняя пустая строка

AlexeyAntipin commented on src/main/java/school/faang/user_service/service/payment/PaymentService.java:
```diff
@@ -0,0 +1,46 @@
+
+    private void validateAndProcessResponse(PaymentRequestDto request, PaymentResponseDto response) {
+        if (!response.getAmount().equals(request.getAmount()) ||
+        !response.getCurrency().toString().equals(request.getCurrency())||
```
Почему везде не оперируешь енамами? Можно же в request тоже получать enum

Raidanmark replied:
не могу разобраться как это делать

AlexeyAntipin replied:
Сделай currency в реквесте не String, а Currency

AlexeyAntipin commented on src/main/java/school/faang/user_service/mapper/TransactionMapper.java:
```diff
@@ -0,0 +1,28 @@
+    TransactionResultDto toDto(Transaction transaction);
+
+    default TransactionStatus mapTransactionStatus(String status) {
+        return TransactionStatus.valueOf(status.toUpperCase());
```
Почему везде не оперируешь енамами? Можно же в dto тоже использовать enum

Raidanmark replied:
не могу разобраться как это делать

AlexeyAntipin commented on src/main/java/school/faang/user_service/service/premium/PremiumService.java:
```diff
@@ -0,0 +1,27 @@
+        PremiumPeriod premiumPeriod = PremiumPeriod.fromDays(durationDays);
+        TransactionResultDto transactionResultDto = transactionService.buyItem(userId, premiumPeriod);
+        if (transactionResultDto.getStatus()
+                .equals(TransactionStatus.SETTLED.toString())) {
```
тут тоже, кажется, везде можно енамы использовать

Raidanmark replied:
не могу разобраться как это делать

AlexeyAntipin replied:
Сделай transactionResultDto поле статус енамом

AlexeyAntipin commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -0,0 +1,18 @@
+
+    public User getUserById(Long id) {
+        return userRepository.findById(id)
+                .orElseThrow(() ->  new EntityNotFoundException("User with id " + id + " not found!"));
```
Используй String.format. Конкатенировать строки - моветон)

AlexeyAntipin commented on src/main/java/school/faang/user_service/UserServiceApplication.java:
```diff
@@ -7,9 +7,12 @@
 
+
 @SpringBootApplication
+@EnableRetry
```
Не увидел в коде использования ретраев. Просто так подключать библиотечку и делать enableRetry не стоит

Raidanmark commented:
Создать таблицу транзакций в бд

Raidanmark committed:
Comment fixes

Raidanmark committed:
Merge branch 'hydra-master-stream10' into BJS2-77432

Raidanmark committed:
Comment fixes

Raidanmark committed:
Comment fixes

Raidanmark committed:
Table Transaction added
Conflicts were resolved

Raidanmark committed:
Test were fixed

Raidanmark committed:
Test were fixed

Raidanmark committed:
Test were fixed

Raidanmark committed:
Test were fixed

Raidanmark committed:
Refactor currency handling to use `Currency` type

Updated various classes and DTOs to use the `Currency` type instead of `String` for better type safety and handling of currency codes. Adjusted related logic, mappers, tests, and imports to ensure compatibility with the refactored type.

Raidanmark committed:
Merge branch 'hydra-master-stream10' into BJS2-77432

# Conflicts:
#	src/main/java/school/faang/user_service/UserServiceApplication.java
#	src/main/java/school/faang/user_service/entity/transaction/Transaction.java
#	src/main/java/school/faang/user_service/entity/transaction/TransactionStatus.java
#	src/main/java/school/faang/user_service/service/skill/SkillService.java
#	src/main/resources/db/changelog/db.changelog-master.yaml
#	src/test/java/school/faang/user_service/service/education/EducationServiceTest.java
#	src/test/java/school/faang/user_service/service/event/EventServiceTest.java

Raidanmark committed:
Standardize table name casing for "transactions"

Updated the "Transactions" table name to "transactions" in both the entity definition and SQL changeset to ensure consistent naming conventions. This improves readability and avoids potential case sensitivity issues in database operations.

Raidanmark committed:
Remove redundant userContext.setUserId call in buyItem

The user ID is already being handled elsewhere and does not need to be set again. This change simplifies the buyItem method and avoids potential redundant operations.

Raidanmark committed:
test fix

AlexeyAntipin commented:
Нужно поправить конфликты

Raidanmark committed:
Merge branch 'hydra-master-stream10' into BJS2-77432
